First try most compatible H5 saving, otherwise use v1.8

oanegros opened the pull request:
fix for #2714, only uses the v1.8 H5 if necessary

k-dominik commented on ilastik/utility/exportFile.py:
```diff
@@ -447,17 +447,25 @@ def write_all(self, mode, compression=None):
-                    )
-                    count += 1
-                    self.ExportProgress(count * 100 / len(self.table_dict))
+            for libver in ["earliest", "v108"]:
```
in principle one could try all values in [`h5py._hl.files.libver_dict`](https://github.com/h5py/h5py/blob/8bed19bed35108190e598993d26bacb70879059c/h5py/_hl/files.py#L35) - do you know of any changes that occurred after `v108` that address the issue of object headers?

Otherwise I would prefer to check first if any of the tables can be written beforehand, but it seems not obvious to determine header size from a structured array.

k-dominik commented on ilastik/utility/exportFile.py:
```diff
@@ -447,17 +447,25 @@ def write_all(self, mode, compression=None):
+                            )
+                            count += 1
+                            self.ExportProgress(count * 100 / len(self.table_dict))
+                except ValueError:
+                    logger.warning("Writing most compatible H5 failed, attempting HDF5 v1.8 format")
```
```suggestion
                except ValueError as e:
                    logger.warning(f"Writing most compatible H5 failed, attempting HDF5 v1.8 format. {e:r}")
```

might be useful to see the actual error here, too - just in case.

k-dominik commented on ilastik/utility/exportFile.py:
```diff
@@ -447,17 +447,25 @@ def write_all(self, mode, compression=None):
+                    logger.warning("Writing most compatible H5 failed, attempting HDF5 v1.8 format")
+                    count = 0
+                    self.ExportProgress(0)
+                    continue
```
It would be nice to check here whether options of writing were exhausted - otherwise this routine will fail "quietly" with a warning and _not_ produce an output.

k-dominik reviewed (changes requested):
Hey @oanegros,

thx a lot for fixing the issue with feature tables with a large number of columns (~2000+). I left a few remarks that would be worth looking into.

Todo for myself: add a test.

oanegros commented:
Added 'latest' hdf5 version to still check if v108 fails to allow future improvements to h5py/HDF5 to allow more than the current v1.8 maximum header size, and added 'loud' failing. 

I don't check for table size before writing as I believe the size of the table header is quite complicated to calculate given datatype conversions.

k-dominik commented on ilastik/utility/exportFile.py:
```diff
@@ -447,17 +447,28 @@ def write_all(self, mode, compression=None):
+                except ValueError as e:
+                    if libver == "latest":
+                        logger.error("latest version of HDF5 export failed in export")
+                        raise ValueError(e)
```
this will generate a "During handling of the above exception, another exception occurred" message. Here it's better to do a bare `raise` (will automatically re-raise `e`).
